src/modules/mc_rate_control/MulticopterRateControl.cpp

```diff
@@ -0,0 +1,412 @@
+			if (_v_control_mode.flag_control_rattitude_enabled) {
+				manual_rate_sp =
+					(fabsf(_manual_control_sp.y) > _param_mc_ratt_th.get()) ||
+					(fabsf(_manual_control_sp.x) > _param_mc_ratt_th.get());
+			}
+
+		} else {
+			_landing_gear_sub.update(&_landing_gear);
+		}
+
+		if (manual_rate_sp) {
+			if (manual_control_updated) {
+
+				// manual rates control - ACRO mode
+				const Vector3f man_rate_sp{
+					math::superexpo(_manual_control_sp.y, _param_mc_acro_expo.get(), _param_mc_acro_supexpo.get()),
+					math::superexpo(-_manual_control_sp.x, _param_mc_acro_expo.get(), _param_mc_acro_supexpo.get()),
+					math::superexpo(_manual_control_sp.r, _param_mc_acro_expo_y.get(), _param_mc_acro_supexpoy.get())};
+
+				_rates_sp = man_rate_sp.emult(_acro_rate_max);
+				_thrust_sp = _manual_control_sp.z;
+
+				// publish rate setpoint
+				vehicle_rates_setpoint_s v_rates_sp{};
+				v_rates_sp.roll = _rates_sp(0);
+				v_rates_sp.pitch = _rates_sp(1);
+				v_rates_sp.yaw = _rates_sp(2);
+				v_rates_sp.thrust_body[0] = 0.0f;
+				v_rates_sp.thrust_body[1] = 0.0f;
+				v_rates_sp.thrust_body[2] = -_thrust_sp;
+				v_rates_sp.timestamp = hrt_absolute_time();
+
+				_v_rates_sp_pub.publish(v_rates_sp);
+			}
+
+		} else {
+			// use rates setpoint topic
+			vehicle_rates_setpoint_s v_rates_sp;
+
+			if (_v_rates_sp_sub.update(&v_rates_sp)) {
+				_rates_sp(0) = v_rates_sp.roll;
+				_rates_sp(1) = v_rates_sp.pitch;
+				_rates_sp(2) = v_rates_sp.yaw;
+				_thrust_sp = -v_rates_sp.thrust_body[2];
+			}
+		}
+
+		// calculate loop update rate while disarmed or at least a few times (updating the filter is expensive)
+		if (!_v_control_mode.flag_armed || (now - _task_start) < 3300000) {
+			_dt_accumulator += dt;
+			++_loop_counter;
+
+			if (_dt_accumulator > 1.0f) {
+				const float loop_update_rate = (float)_loop_counter / _dt_accumulator;
+				_loop_update_rate_hz = _loop_update_rate_hz * 0.5f + loop_update_rate * 0.5f;
+				_dt_accumulator = 0;
+				_loop_counter = 0;
+				_rate_control.setDTermCutoff(_loop_update_rate_hz, _param_mc_dterm_cutoff.get(), true);
+			}
+		}
+
+		// run the rate controller
+		if (_v_control_mode.flag_control_rates_enabled && !_actuators_0_circuit_breaker_enabled) {
+
+			// reset integral if disarmed
+			if (!_v_control_mode.flag_armed || _vehicle_status.vehicle_type != vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
+				_rate_control.resetIntegral();
+			}
+
+			// update saturation status from mixer feedback
+			if (_motor_limits_sub.updated()) {
+				multirotor_motor_limits_s motor_limits;
+
+				if (_motor_limits_sub.copy(&motor_limits)) {
+					MultirotorMixer::saturation_status saturation_status;
+					saturation_status.value = motor_limits.saturation_status;
+
+					_rate_control.setSaturationStatus(saturation_status);
+				}
+			}
+
+			// run rate controller
+			const Vector3f att_control = _rate_control.update(rates, _rates_sp, dt, _maybe_landed || _landed);
```

Comment:
In the previous implementation, the rate control was done at the very beginning to minimize control latency.
Here, there are a few things that can introduce delay between gyro update and publication of actuator_controls : updates of low pass filters, a few uORB copies/pubs for manual control, etc. 

Have you

Reply:
Yes, although that's already somewhat the case and something to be aware of beyond this PR. For the inner loop (wq:rate_ctrl) it's expected these modules run each cycle to completion back to back. 

The manual handling can be further optimized to reduce unnecessary work.

Reply:
Oh yes you are right, this is running in a work queue.

Good reason to make things more granular. I was considering to pull manual setpoint into a separate task, publishing setpoints based on the current mode and airframe type.

Reply:
> I was considering to pull manual setpoint into a separate task

Likewise, and I even started sketching it out. Where I got stuck was the current rattitude handling.

Reply:
Otherwise we could almost think of this as immediately splitting mc_att_control into 3 pieces.

 - mc_rate_control
 - mc_att_control
 - mc_stick_mapper (or better name)

Reply:
The landing gear is also a bit of a mess. It's pretty clear to me now we shouldn't have lanig gear in actuator_controls_0.

https://github.com/PX4/Firmware/blob/9e309f62a9b1731caae96000b824aa96661e67ad/msg/actuator_controls.msg#L11

Reply:
> immediately splitting mc_att_control into 3 pieces.

Sounds good to do it incrementally. It can be done right after this PR. Then, the same for FW, vtol, etc.
Whether it makes sense to merge mc_stick_mapper, fw_stick_mapper, vtol_stick_mapper and rover_stick_mapper will be obvious once they exist.

Reply:
>  we shouldn't have landing gear in actuator_controls_0

Don't tell me. "Easy" fix is to move it to another control group and update all mixers

Reply:
> we shouldn't have landing gear in actuator_controls_0

Totally, it was even in the attitude setpoint which @simonegu luckily removed in a refactoring effort. I think not having it in the actuator control is the next step.

Reply:
> mc_stick_mapper (or better name)
> based on control mode publishes attitude or rate setpoints based on manual_control_setpoint input.

The acro setpoint generation is quite simple but I think there's a lot of reusability potential even with altitude, position mode because currently the position controller generates thrust (soon acceleration) from stick this gets translated into a full attitude setpoint and then torn apart again by the attitude controller just because of the interface. I plan to improve the interface to allow less back and forth.